### PR TITLE
fix: RequestEventAction has missed a type `resolveValue`

### LINKED ISSUE #7377: [🐞] "resolveValue" Function Is Missing in TypeScript Types and Fails in "routeAction$" / "globalAction$"
### Which component is affected?

Qwik City (routing)

### Describe the bug

The [resolveValue](https://qwik.dev/docs/route-loader/#access-the-routeloader-data-within-another-routeloader) function is documented as being available for `routeLoader$`, `routeAction$`, and `globalAction$`. However, there are two major issues:  

1. **Missing TypeScript Typings**  
   - While `resolveValue` exists in the `requestEvent` of `routeAction$` and `globalAction$`, it is missing from TypeScript types. This results in TypeScript failing to recognize it, leading to type errors.  

2. **Functionality Not Working in Actions**  
   - When attempting to use `resolveValue` in `routeAction$` or `globalAction$`, it throws an error instead of resolving the expected value.  

### **Expected Behavior**  
The code should work similarly to `routeLoader$`, allowing the resolution of a loader's value within an action.

### **Actual Behavior**  
The following error is thrown:  

```
[vite] Internal server error: You can not get the returned data of a loader that has not been executed for this request.
```

This suggests that `resolveValue` is either not supported or has unintended behavior when used in `routeAction$` or `globalAction$`.

### Reproduction

https://stackblitz.com/edit/github-4169k7qj?file=src%2Froutes%2Findex.tsx

### **Steps to Reproduce**  
Using `resolveValue` inside a `routeAction$` function:  

```ts
export const useAction = routeAction$(
  async (data, request) => {
    const user = await request.resolveValue(useUser);
    console.log(user);
  }
);
```

### System Info

```shell
System:
    OS: Windows 11 10.0.26100
    CPU: (8) x64 12th Gen Intel(R) Core(TM) i3-12100F
    Memory: 3.85 GB / 15.82 GB
  Binaries:
    Node: 23.6.0 - C:\Program Files\nodejs\node.EXE
    npm: 11.1.0 - C:\Program Files\nodejs\npm.CMD
    pnpm: 9.15.3 - C:\Program Files\nodejs\pnpm.CMD
  Browsers:
    Edge: Chromium (131.0.2903.112)
    Internet Explorer: 11.0.26100.1882
  npmPackages:
    @builder.io/qwik: ^1.12.0 => 1.12.0
    @builder.io/qwik-city: ^1.12.0 => 1.12.0
    vite: 6.1.0 => 6.1.0
```

### Additional Information

Being able to access `routeLoader$` values inside `routeAction$` and `globalAction$` is highly beneficial, as it allows reusing computed loader data without needing to fetch or compute it again.

Would it be possible to officially support this functionality, or is this an unintended limitation? Additionally, the TypeScript types should be updated to reflect `resolveValue`'s presence in `requestEvent`.

### PR DESCRIPTION
**Pull request by JerryWu1234**
# What is it?

fixes #7377 

if routeAction$ wants to use `resolveValue` function，routeLoader$ has be loaded first, 
so I move the `routerLoader` logic before `method === 'POST'`, However, in some situations, it has to remain in its original position.

- Feature / enhancement
- Bug
- Docs / tests / types / typos
- Infra

# Description

# Checklist

- [x] My code follows the [developer guidelines of this project](https://github.com/QwikDev/qwik/blob/main/CONTRIBUTING.md)
- [x] I performed a self-review of my own code
- [x] I added a changeset with `pnpm change`
- [x] I made corresponding changes to the Qwik docs
- [x] I added new tests to cover the fix / functionality

### CODE REVIEW

**Comment by wmertens** [comment]
the problem is that the resolveValue is not actually working even though it's documented. So this PR needs to implement it too.

**Comment by JerryWu1234** [comment]
> the problem is that the resolveValue is not actually working even though it's documented. So this PR needs to implement it too.

Is there a code reference I could follow ?

**Comment by wmertens** [comment]
@JerryWu1234 best to search the code for `resolveValue` and see how it is implemented, and then try to implement it in the same way in the other locations.

**Comment by JerryWu1234** [comment]
> @JerryWu1234 best to search the code for `resolveValue` and see how it is implemented, and then try to implement it in the same way in the other locations.

thanks for your tips, it's useful

**Comment by JerryWu1234** [comment]
@wmertens @Varixo please help to review it when you are available

File: packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts [hunk 1]
```diff
@@ -257,9 +227,49 @@ export function actionsMiddleware(routeActions: ActionInternal[], routeLoaders:
...
+            loaders[selectedActionId] = actionResolved;
+            isAction = true;
+          }
+        }
+      }
+    }
+    if (routeLoaders.length > 0 && isAction) {
+      await Promise.all(loader() || []);
     }
+    isAction = false;
```
Comment (gioboa) [inline packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts:272]: Do we need to reset it to false?

File: packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts [hunk 2]
```diff
@@ -257,9 +227,49 @@ export function actionsMiddleware(routeActions: ActionInternal[], routeLoaders:
...
+            }
+
+            loaders[selectedActionId] = actionResolved;
+            isAction = true;
+          }
+        }
+      }
+    }
+    if (routeLoaders.length > 0 && isAction) {
+      await Promise.all(loader() || []);
```
Comment (gioboa) [inline packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts:270]: we already run loaders at line 189 :thinking: is it correct?

File: packages/qwik-city/src/middleware/request-handler/request-event.ts [hunk 3]
```diff
@@ -280,7 +279,7 @@ export function createRequestEvent(
       return writableStream;
     },
   };
-  return Object.freeze(requestEv);
+  return requestEv;
```
Comment (gioboa) [inline packages/qwik-city/src/middleware/request-handler/request-event.ts:282]: Can you explain me the `Object.freeze` removal please?

**Review by gioboa — CHANGES_REQUESTED** [review]
Thanks @JerryWu1234 for your help :tada: 
I added few comments

File: packages/qwik-city/src/middleware/request-handler/request-event.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (wmertens) [reply packages/qwik-city/src/middleware/request-handler/request-event.ts:282]: @JerryWu1234 please put this back

File: packages/qwik/src/optimizer/src/qwik-binding-map.ts
Comment (wmertens) [inline packages/qwik/src/optimizer/src/qwik-binding-map.ts:1]: please remove these changes from the PR (run `git checkout main -- packages/qwik/src/optimizer/src/qwik-binding-map.ts`)

File: packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts [hunk 4]
```diff
@@ -184,43 +185,12 @@ export function actionsMiddleware(routeActions: ActionInternal[], routeLoaders:
...
-            }
-            loaders[selectedActionId] = actionResolved;
-          }
-        }
-      }
-    }
 
     if (routeLoaders.length > 0) {
-      const resolvedLoadersPromises = routeLoaders.map((loader) => {
+      await Promise.all(loader() || []);
```
Comment (wmertens) [inline packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts:190]: I'd prefer `await Promise.all(routeLoaders.map((loader) => ... (previous code) )`

File: .changeset/eight-garlics-deliver.md [hunk 5]
```diff
@@ -0,0 +1,6 @@
+---
+'@builder.io/qwik-city': patch
+'@builder.io/qwik': patch
+---
+
+FIX: if routeAction$ wants to use resolveValue function，routeLoader$ has be loaded first,
```
Comment (wmertens) [inline .changeset/eight-garlics-deliver.md:6]: I don't understand how this PR implements what you write. How about this?
```
FIX: add type for `routeAction$`'s `requestEvent.resolveValue`
```

**Review by wmertens — CHANGES_REQUESTED** [review]

File: packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts (hunk 2 above)
Comment (JerryWu1234) [reply packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts]: if routeAction$ wants to use resolveValue function，routeLoader$ has be loaded first,
so I move the routerLoader logic before method === 'POST', However, in some situations, it has to remain in its original position.

File: packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts (hunk 1 above)
Comment (JerryWu1234) [reply packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts:270]: We need to prevent it from running multiple times.

File: .changeset/eight-garlics-deliver.md [hunk 6]
```diff
@@ -0,0 +1,6 @@
+---
+'@builder.io/qwik-city': patch
+'@builder.io/qwik': patch
```
Comment (Varixo) [inline .changeset/eight-garlics-deliver.md:3]: I think this is qwik-city only change

File: packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts [hunk 7]
```diff
@@ -184,6 +185,48 @@ export function actionsMiddleware(routeActions: ActionInternal[], routeLoaders:
         );
       }
     }
+
+    if (routeLoaders.length > 0) {
+      await Promise.all(routeLoaders.map(loader));
+    }
+    function loader(loader: LoaderInternal) {
```
Comment (wmertens) [inline packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts]: please put this inside the map function, it's confusing to have it outside and after, especially with the name `loader` that you also shadow inside it

File: packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts [hunk 8]
```diff
@@ -213,53 +257,17 @@ export function actionsMiddleware(routeActions: ActionInternal[], routeLoaders:
...
-            }
-            return resolvedLoader;
-          });
-
-        return loaders[loaderId];
-      });
-
-      await Promise.all(resolvedLoadersPromises);
+    if (routeLoaders.length > 0 && isAction) {
+      await Promise.all(routeLoaders.map(loader));
```
Comment (wmertens) [inline packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts]: I don't understand why we run the routeloaders again? Seems wasteful?
(but now I understand why you moved it to a separate function)

File: packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts (hunk 1 above)
Comment (wmertens) [reply packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts]: how does that happen?

File: packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts (hunk 7 above)
Comment (JerryWu1234) [reply packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts]: > please put this inside the map function, it's confusing to have it outside and after, especially with the name `loader` that you also shadow inside it

loader be used by two place.
I can't put this inside map.

File: packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts (hunk 8 above)
Comment (JerryWu1234) [reply packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts]: > I don't understand why we run the routeloaders again? Seems wasteful? (but now I understand why you moved it to a separate function)

in some edge cases.  routeLoader function has to run again.
Before I just moved that code before POST, but There is still some fail tests.

File: packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts (hunk 7 above)
Comment (JerryWu1234) [reply packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts]: @wmertens I changed that name for clarify

**Comment by wmertens** [comment]
@JerryWu1234 so can you explain the fix? You now run the routeLoaders before the action, but then you run them again after the action, why?

**Comment by wmertens** [comment]
Ah because the action can change the loaders, right? 

In that case, I think it's better to make the first routeLoaders run on demand. We should not add the resolveValue synchronous API, but instead an async call like getLoaderValue(loader).

Right now, we're doubling the cost of loaders in actions in this PR

**Comment by JerryWu1234** [comment]
> Ah because the action can change the loaders, right? 
> 
> 
> 
> In that case, I think it's better to make the first routeLoaders run on demand. We should not add the resolveValue synchronous API, but instead an async call like getLoaderValue(loader).
> 
> 
> 
> Right now, we're doubling the cost of loaders in actions in this PR

It's a long time PR, so I  don't remember very clearly. Let me re-read the code tomorrow.

**Comment by JerryWu1234** [comment]
> Ah because the action can change the loaders, right?
> 
> In that case, I think it's better to make the first routeLoaders run on demand. We should not add the resolveValue synchronous API, but instead an async call like getLoaderValue(loader).
> 
> Right now, we're doubling the cost of loaders in actions in this PR

```js

export const useAddUser = routeAction$(
  async (data, requestEvent) => {
    const res = await requestEvent.resolveValue(useUser);
    res.firstName = data.firstName;
    res.lastName = data.lastName;
    return {
      success: true,
      userID: res,
    };
  },
  zod$({
    firstName: z.string(),
    lastName: z.string(),
  }),
);
```

you meant that user has to use `getLoaderValue ` instead of `resolveValue`?
@wmertens

File: packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts [hunk 9]
```diff
@@ -268,6 +234,41 @@ export function actionsMiddleware(routeActions: ActionInternal[], routeLoaders:
 
       await Promise.all(resolvedLoadersPromises);
     }
+
+    if (method === 'POST') {
```
Comment (gioboa) [inline packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts:238]: moved the POST section after the loaders resolution.

**Comment by JerryWu1234** [comment]
@gioboa 
image

image

https://github.com/JerryWu1234/qwik/blob/377cd5dc92cfdc6daa3cf3a3f3c7d80b7c730b8c/packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts#L267



There are some `LoaderInternal`s still needs to be loaded after POST.

So Why I executed `routeLoaders` twice

**Comment by wmertens** [comment]
hmm looks like the e2e tests actually aren't passing, I tried 3 times and normally it's only the macos one that is slow

**Comment by JerryWu1234** [comment]
> hmm looks like the e2e tests actually aren't passing, I tried 3 times and normally it's only the macos one that is slow

this method isn't fully fix this problem.
I tried my local. the wrong is I wrote e2e that didn't pass .

**Comment by gioboa** [comment]
Nice @JerryWu1234 
We don't want to run loaders twice, before and after because the call must be unique.
Immagine a db in the loader, should run only one time. Why is it failing the test? That's the question I guess

**Comment by JerryWu1234** [comment]
> Why is it failing the test? That's the question I guess

```js
export const globalAction = globalAction$(
  async (data, requestEvent) => {
    const res = await requestEvent.resolveValue(useUserGlobal);
    if (res) {
      res.firstName = data.globalfirstName;
      res.lastName = data.globallastName;
    }
    return {
      success: true,
      userID: res,
    };
  },
  zod$({
    globalfirstName: z.string(),
    globallastName: z.string(),
  }),
);
```

because in globalAction function that involved `routerloader`.

**Comment by gioboa** [comment]
I see, we already executed the loader though 🤔
